packages/type-safe-api/src/construct/integrations/s3.ts

```diff
@@ -0,0 +1,179 @@
+
+  constructor(props: S3IntegrationProps) {
+    super();
+
+    this.bucket = props.bucket;
+
+    if (props.role) {
+      this.role = props.role;
+    } else {
+      this.role = (this.bucket.node.tryFindChild("ExecutionRole") ??
+        new Role(this.bucket, "ExecutionRole", {
+          assumedBy: new CompositePrincipal(),
+        })) as IRole;
+    }
+
+    this.method = props.method;
+    this.path = props.path;
+    this.errorIntegrationResponse = props.errorIntegrationResponse;
+  }
+
+  /**
+   * Render the S3 integration as a snippet of OpenAPI
+   */
+  public render(props: IntegrationRenderProps): ApiGatewayIntegration {
+    return {
+      type: "AWS",
+      httpMethod: (this.method ?? props.method).toUpperCase(),
+      uri: bucketInvocationUri(this.bucket, this.path ?? props.path),
+      credentials: this.role.roleArn,
+      responses: {
+        default: {
+          statusCode: "200",
+          responseParameters: props.corsOptions
+            ? generateCorsResponseParameters(props.corsOptions)
+            : {},
+          responseTemplates: {},
+        },
+
+        ...(
+          this.errorIntegrationResponse ?? ErrorIntegrationResponses.catchAll()
+        ).render(props),
+      },
+    };
+  }
+
+  /**
+   * Grant API Gateway permissions to invoke the S3 bucket
+   */
+  public grant({ scope, api, method, path }: IntegrationGrantProps) {
+    this.role.grant(
+      new ServicePrincipal("apigateway.amazonaws.com", {
+        conditions: {
+          "ForAnyValue:StringEquals": {
+            "aws.SourceArn": Stack.of(scope).formatArn({
+              service: "execute-api",
+              resource: api.restApiId,
+              // Scope permissions to any stage and a specific method and path of the operation.
+              // Path parameters (eg {param} are replaced with wildcards)
+              resourceName: `*/${method.toUpperCase()}${path.replace(
+                /{[^\}]*\}/g,
+                "*"
+              )}`,
+            }),
+          },
+        },
+      })
+    );
+
+    const bucketMethod = this.method ?? method;
+
+    var bucketActions: string[] = [];
+    var keyActions: string[] = [];
+
+    switch (bucketMethod) {
+      case "get":
+      case "head":
+      case "options":
+      case "trace":
+        bucketActions = perms.BUCKET_READ_ACTIONS;
+        keyActions = perms.KEY_READ_ACTIONS;
+        break;
+
+      case "post":
+      case "put":
+      case "patch":
+        bucketActions = FeatureFlags.of(scope).isEnabled(
+          cxapi.S3_GRANT_WRITE_WITHOUT_ACL
+        )
+          ? perms.BUCKET_PUT_ACTIONS
+          : perms.LEGACY_BUCKET_PUT_ACTIONS;
+        keyActions = perms.KEY_WRITE_ACTIONS;
+        break;
+
+      case "delete":
+        bucketActions = perms.BUCKET_DELETE_ACTIONS;
+        break;
+    }
+
+    Grant.addToPrincipalOrResource({
+      grantee: this.role,
+      actions: bucketActions,
```

Comment:
Nice idea having the grant shared but switching on the method for more minimal permissions :)

Reply:
So you prefer this more complex approach ? I can keep it?

Reply:
Yeah I think let's keep it :)